### PR TITLE
Duplicate Balancing Transformer

### PR DESCRIPTION
**Pull request by rbharath**
First step towards new variant of balancing transformer

### CODE REVIEW

**Comment by rbharath** [comment]
This PR ended up being a little more complicated than I thought. The primary reason for the complexity is that `DuplicateBalancingTransformer` actually changes the number of datapoints in a dataset as part of a tranformation. This is because `DuplicateBalancingTransformer` is closer to a data augmentation technique than the transformations we've implemented thus far. As a result, I had to change the signature of `Transformer.transform_array` to allow `ids` to be altered as well. This required additional changes in `dc.data.datasets` as well.

In addition to these changes I've made a few more additions:

- Added in example doctests for almost all the transformers (with the exception of the broken `ANITransformer` #2047)
- Separated out tests for various transformers into separate files
- Started adding in type hints for transformers, although I haven't finished the task in this PR.

CC @peastman @nd-02110114 @ncfrey

File: deepchem/trans/duplicate.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+from deepchem.trans.transformers import Transformer
+
+logger = logging.getLogger(__name__)
+
+
+class DuplicateBalancingTransformer(Transformer):
+  """Balance binary or multiclass datasets by duplicating rarer class samples.
+
+  This class balances a dataset by duplicating samples of the rarer class so
+  that the sum of all example weights from all classes is the same. (Up to
```
Comment (peastman) [inline deepchem/trans/duplicate.py:14]: I don't think that's what it actually does.  To tries to balance the *number* of samples in each class, not the *sum of weights* for each class.  Although possibly the latter is what it really should do?

Comment (rbharath) [reply deepchem/trans/duplicate.py:14]: Ah good catch! I agree that the latter should be the expected behavior. I've fixed this and added an additional test for weight handling

Comment (peastman) [reply deepchem/trans/duplicate.py:14]: Looks good!

**Comment by rbharath** [comment]
Looks like the only failure is test_roulette so going to go ahead and merge!

**Pull request merged**